### ISSUE TITLE
Loss is zero or nan during training on another dataset.

### ISSUE THREAD

**Issue by warmspringwinds** [s9xie/hed#12]
Hello.

While training your network on different dataset I get  zero loss or nan after first couple of iterations.

Do you know why it happens?

I also can't get how the ground truth in your dataset is marked: I notice values between 0 and 255.
Bigger number means stronger edge?
Is it ok to use this labels with sigmoid  cross entropy loss?

How should I label my own dataset?

Thank you.

**Comment by extragoya** [comment]
I also get this problem with my own dataset. The only way I could get the loss to stop exploding was to use learning rates of 1*10e-8, but I feel like this is getting way too small...I'd be interested to know what other people had to do to get this working on their other datasets.

**Comment by warmspringwinds** [comment]
@extragoya I also solved the problem by decreasing the learning rate.
Seemed to make it work.

**Comment by zeakey** [comment]
smaller learning rate may be helps

**Comment by cchenzhou** [comment]
hi, I also to use learning rates of 1*10e -8,but the value of loss id very big even Twenty thousand.what is your learning rate?thank you.

**Comment by codecolony** [comment]
@cchenzhou  - Don't worry about bigger loss value. It means nothing from my experience. The results should show up as expected.

**Comment by cchenzhou** [comment]
@codecolony thank you very much.When the iteration equals 100000 the training is stop,but the loss value equals 22758.3.I can understand why.Any idea about how to fix the error?

**Comment by brisker** [comment]
@cchenzhou   @warmspringwinds   @extragoya   @zeakey    @codecolony 
Have anybody met this problem when at the very start of training , when running "solve.py" ?
Any advice is appreciated, thanks !
I0306 11:33:13.509395  3994 layer_factory.hpp:76] Creating layer data
I0306 11:33:13.509449  3994 net.cpp:111] Creating Layer data
I0306 11:33:13.509469  3994 net.cpp:434] data -> data
I0306 11:33:13.509498  3994 net.cpp:434] data -> label
I0306 11:33:13.509524  3994 image_labelmap_data_layer.cpp:40] Opening file ../../data/HED-BSDS/train_pair.lst
I0306 11:33:13.535454  3994 image_labelmap_data_layer.cpp:50] Shuffling data
I0306 11:33:13.540946  3994 image_labelmap_data_layer.cpp:55] A total of 28800 images.
libpng warning: Application built with libpng-1.2.54 but running with 1.6.27
E0306 11:33:13.545475  3994 io.cpp:77] Could not open or find file ../../data/HED-BSDS/train/aug_gt_scale_0.5/157.5_1_0/159045.png
F0306 11:33:13.545496  3994 image_labelmap_data_layer.cpp:81] Check failed: (height == gt_height) && (width == gt_width) groundtruth size != image size
*** Check failure stack trace: ***

**Comment by zeakey** [comment]
@brisker It clearly tells you your ground-truth size doesn't match corresponding image size.

**Comment by brisker** [comment]
@zeakey 
I have checked that I do have that image in my dataset folder

**Comment by extragoya** [comment]
@brisker Notice the error says: "Could not open or find file ../../data/HED-BSDS/train/aug_gt_scale_0.5/157.5_1_0/159045.png" - the problem is it can't find an image in your training list. The image_labelmap_data_layer will keep running, however, and then throw the final error when the sizes don't match. I would avoid the relative paths and make sure there are no spaces or anything else funny in your training list.

**Comment by brisker** [comment]
@zeakey @extragoya
I have checked that I do have that image in my dataset folder.
The dataset is just downloaded from the link provided by the author, I really can not understand why this happens! Besides, I have replaced the path to absolute path, no help..

**Comment by extragoya** [comment]
@brisker For one reason or the other, it's unable to open the file. I would assume that something is wrong with the path - what do you have as root_folder under image_data_param in your train_val.prototxt?

**Comment by brisker** [comment]
@extragoya 
At the begining, it was "../../data/HED-BSDS/", and error occurs.
Then, I replace it with "/home/jcc/code/hed/data/HED-BSDS/"
Still error...

**Comment by brisker** [comment]
@extragoya 
I  think the error clearly shows that the code has found the train_pair.lst, so why can not find the images?

**Comment by extragoya** [comment]
@brisker The path to the train_pair.lst is given in your train_val prototx file, whereas the paths to the images are given in train_pair.lst and your root_folder - so the path to the list can be correct whereas the paths to the images could be incorrect. Also, train_pair.lst is a text file, so it is opened differently than an image. Do you have opencv installed, and did you turn opencv on in your makefile.config?

**Comment by zeakey** [comment]
@brisker 
Could you plz give me the situation about the directories, including where you run `solve.py`(by default you should run `solve.py` in `hed/examples/hed/`), and where you put your dataset, and an example line of your train_pair.lst.

Further more, please check the permission of the dataset, do you have reading permision ?

**Comment by brisker** [comment]
@extragoya 
I have opencv 3 installed, but I can not see any flags indicating whether turning opencv on in the  makefile.config here: https://github.com/s9xie/hed/blob/master/Makefile.config.example   I think I got the opencv well compiled with caffe.

**Comment by zeakey** [comment]
@cchenzhou In my experience the loss is of little impact, the loss value is still big even model reach convergence. I suggest you to test your model on validation set. Do you use your own dataset? NOTE the positive/negative ration because  SoftmaxLoss in HED will count adjust the positive/negative ratio.

**Comment by brisker** [comment]
@zeakey 
I run solve.py in hed/examples/hed/. and the data  folder is in /home/jcc/code/hed ,just the root folder.
The line of train.lst seems like  train/aug_gt_scale_0.5/157.5_1_0/159045.png

**Comment by zeakey** [comment]
@brisker OK, seems everything is just right...
You just remember **the `root_folder` you specify in ImageLabelmapDataLayer concat the item in train_pair.lst is where the image/gt file locates**.

**Comment by extragoya** [comment]
@brisker Ok, later versions of caffe have an option to turn OPEN_CV off and also to specify what version you're using: https://github.com/BVLC/caffe/blob/master/Makefile.config.example. It may not apply with HED's version of caffe, but that is why I asked.

**Comment by brisker** [comment]
@extragoya  @zeakey 
Thanks a lot for your replies!
No other advice?

**Comment by zeakey** [comment]
Set up breakpoint and debug into the code to see details.
在 2017年3月7日，上午12:08，numberjcc <notifications@github.com<mailto:notifications@github.com>> 写道：


@extragoya<https://github.com/extragoya> @zeakey<https://github.com/zeakey>
Thanks a lot for your replies!
No other advice?

**Comment by extragoya** [comment]
@zeakey @brisker Agreed. I would add that you should try to determine if the problem is whether the code cannot find the image, or can find the image but cannot load it.

**Comment by cchenzhou** [comment]
@zeakey I have been tested my model on test set，but the result is very bad.I can't see any contour on the output image,so I think my model don't reach convergence.I use their BSDS500 and only change the learning rate.Could you tell me any parameters should be change or adjust the positive/negative ratio?Thank you!

**Comment by brisker** [comment]
@extragoya @zeakey 
I reinstalled the opencv and get the problem solved. Besides, could you please tell me where to fine a toolbox for the performance evaluation of the edge detection results? Thanks!

**Comment by brisker** [comment]
@zeakey 
@extragoya 
#28 
I also want to know what is the relationship among the printed losses, any idea?
Besides, it seems that sigmoidCrossEntropy loss should not be as larger as several thousand, why the printed losses values are so big even when the model has converged?
![image](https://cloud.githubusercontent.com/assets/13804492/24072493/b7331d4a-0c22-11e7-8342-079029157c02.png)

**Comment by codecolony** [comment]
@cchenzhou - What is the learning rate you're using?

**Comment by brisker** [comment]
@codecolony 
Hi, the modified SigmoidCrossEntropy Loss layer has a [line](https://github.com/s9xie/hed/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cpp#L107)
that reads like -:

    bottom_diff[i * dim + j] *= 1 * count_neg / (count_pos + count_neg);

why not 

     bottom_diff[i * dim + j] = 1 * count_neg / (count_pos + count_neg);
?
It seems that the gradients are multiplied in the loops. It is a little confusing to me.  Why does the author write like that?

**Comment by dxytz** [comment]
@cchenzhou 
Hi,
My model can't reach convergence too. How about yours? My label map is a binary image(0-255) with one channel. Do you have any advice? 
Tank you!
